### ISSUE TITLE
scheduler: ambiguity among feature gates, custom scheduler policy and internal predicates/priorities

### ISSUE THREAD

**Issue by Huang-Wei** [kubernetes/kubernetes#80143]
How the scheduler behaves depends on (a) enabled feature gates and (b) registered predicates/priorities, the combinations could be:

- feature gates + default policy (i.e. default predicates/priorities)
- feature gates + custom policy (i.e. custom predicates/priorities)

But I see some issues around the consistency among feature gates, custom policy and internal state of scheduler predicate/priority. Put here for brainstorming:

## 1. Invalid custom scheduler policy
	
A predicate/priority may depend on or conflict with others. If the predicates/priorities specified in a custom policy file are invalid, schedule may not function well.
	
**Solution:** needs a validation mechanism to abort scheduler startup - #79426

## 2. Additional predicates/priorities are appended on top of custom scheduler policy
	
Internally upon startup of scheduler, two parts of predicates/priorities are applied on top of the custom ones (loaded from custom policy file):
	
1. MandatoryPredicate (see `RegisterMandatoryFitPredicate()`)
1. Modified Predicates/Priorities in `ApplyFeatureGates()`

The MandatoryPredicate behavior looks good to me if it's truly mandatory for generic scheduling purpose. But the latter behavior concerns me as it may alter the user input vastly.

**Solution:** need to discuss which is expected behavior; if appropriate, skip running `ApplyFeatureGates()` if a custom scheduler policy is given.

## 3. A reasonable state of default predicates/priorities

`defaultPredicates()` and `defaultPriorities()` represents the default state of scheduler's functionalities. But at this moment, the default ones doesn't follow a specific pattern. We'd make it very clear - whether it's (a) GAed and Beta pred/prio are default, or (b) just GAed pred/prio are default. Only following a clear pattern, the system can be stable along with rollout of individual feature.

**Solution:** discuss the pattern first and revisit default predicates and priorities.

## 4. `ApplyFeatureGates()` is not stateless

We can't assume `ApplyFeatureGates()` is only called in a real cluster or e2e test. It can be called in integration tests with case (a) disabled beta features and (b) enabled alpha features, or worse: combined with the aforementioned ambiguous pattern.

Here is an example:

1. TestSchedulerCreationFromConfigMap (in test/integration/scheduler/scheduler_test.go) doesn't call `ApplyFeatureGates()`, and expect a mandatory predicate to be "CheckNodeConditionPred"
1. TestTaintNodeByCondition (in test/integration/scheduler/taint_test.go) calls `ApplyFeatureGates()`, and expect the mandatory predicate to be "PodToleratesNodeTaintsPred" (which replaces "CheckNodeConditionPred")

Apparently, the latter test **mutates** the default state without setting it back, and the first one rely on the original default state. So if the latter one runs prior to first one, it'd get failed. However, the file `scheduler_test.go` is alphabetically ahead of `taint_test.go`, so in Golang, tests in `scheduler_test.go` runs prior to tests in `taint_test.go`, so we haven't see the "potential" issue so far. But we can make a test to trigger it:

```
$ mv test/integration/scheduler/taint_test.go test/integration/scheduler/a_test.go
$ make test-integration WHAT=./test/integration/scheduler KUBE_TEST_ARGS="-run ^Test\(TaintNodeByCondition\|SchedulerCreationFromConfigMap\)"
......
--- FAIL: TestSchedulerCreationFromConfigMap (4.09s)
    scheduler_test.go:291: Expected predicates map[CheckNodeCondition:{} PredicateOne:{} PredicateTwo:{}], got map[CheckNodeUnschedulable:{} PodToleratesNodeTaints:{} PredicateOne:{} PredicateTwo:{}]
......
```

**Solution:** make `ApplyFeatureGates()` stateless - #80144

**Comment by Huang-Wei** [comment]
/sig scheduling
cc/ @bsalamat @k82cn @ravisantoshgudimetla @ahg-g

**Comment by ahg-g** [comment]
/remove-lifecycle stale

Most of the concerns in this issue can be addressed as part of phase 2 of the framework migration.

**Comment by ahg-g** [comment]
I think this can be closed

**Comment by Huang-Wei** [comment]
1 is still valid, we should come with validation against the CC (ComponentConfig), but only on in-tree plugins. For example, it doens't make sense to enable both MostRequested and LeastRequested score plugins at the same time. We can open an issue later.

2, 3, 4 are resolved due to the refactoring in 1.18. Now we don't have global predicate/priority keys, and plugins are managed by the framework instance, which is convenient now to build stateless tests.
Due to that, featuregate won't leave enabled (or disabled) plugin in the wild - it's within a framework instance, not globally. As the framework instance lives within a test (or subtest), it's unnecessary to unregister the plugin.
